File: src/athena/research/supervisor/policy.py
"""Replaceable hypothesis scheduling policy boundary."""

import math
from enum import StrEnum
from typing import Protocol

from athena.core.research_models import Hypothesis
from athena.core.research_tree import ResearchTree


class Outcome(StrEnum):
    """Trusted result of a candidate against its frozen reference."""

    WIN = "WIN"
    DRAW = "DRAW"
    LOSS = "LOSS"


class HypothesisPolicy(Protocol):
    """Narrow policy used by the Supervisor scheduler."""

    def seed(self, parent: Hypothesis | None) -> float: ...

    def priority(self, hypothesis: Hypothesis, tree: ResearchTree) -> float: ...

    def settle(self, reference_priority: float, outcome: Outcome) -> float: ...


# TODO(search-policy): Replace EloPolicy with an evidence-aware scheduling
# policy after enough real-search traces exist. Keep Supervisor dependent only
# on HypothesisPolicy.
class EloPolicy:
    """Single-sided Elo update with a fixed 0.5 expected score."""

    ROOT_PRIORITY = 1000.0

    def __init__(self, *, k: float = 32.0) -> None:
        if not math.isfinite(k) or k <= 0:
            raise ValueError("Elo k must be a positive finite number")
        self._k = k

    def seed(self, parent: Hypothesis | None) -> float:
        return self.ROOT_PRIORITY if parent is None else parent.priority

    def priority(self, hypothesis: Hypothesis, tree: ResearchTree) -> float:
        del tree
        return hypothesis.priority

    def settle(self, reference_priority: float, outcome: Outcome) -> float:
        if not math.isfinite(reference_priority):
            raise ValueError("reference priority must be finite")
        score = {
            Outcome.WIN: 1.0,
            Outcome.DRAW: 0.5,
            Outcome.LOSS: 0.0,
        }[outcome]
        return reference_priority + self._k * (score - 0.5)


def queue_order(priority: float, order: int | None) -> tuple[float, int]:
    """Sort key for priority descending followed by FIFO order ascending."""
    if not math.isfinite(priority):
        raise ValueError("hypothesis priority must be finite")
    if order is None:
        raise ValueError("hypothesis order must be assigned before scheduling")
    return (-priority, order)

File: athena_ts/packages/athena-research/src/supervisor/policy.ts
/**
 * 可替换的假设调度策略边界（移植 ``research/supervisor/policy.py``）。
 */

import type { Hypothesis, ResearchTree } from "@athena/core"

/** 候选相对其冻结参照的可信结果。 */
export const Outcome = {
  WIN: "WIN",
  DRAW: "DRAW",
  LOSS: "LOSS",
} as const
export type Outcome = (typeof Outcome)[keyof typeof Outcome]

/** Supervisor 调度器使用的窄策略。 */
export interface HypothesisPolicy {
  seed(parent: Hypothesis | null): number
  priority(hypothesis: Hypothesis, tree: ResearchTree): number
  settle(referencePriority: number, outcome: Outcome): number
}

/**
 * 单侧 Elo 更新，固定 0.5 期望得分。
 *
 * TODO(search-policy): 有足够真实搜索轨迹后，用 evidence-aware 调度策略替换
 * EloPolicy；Supervisor 只依赖 HypothesisPolicy。
 */
export class EloPolicy implements HypothesisPolicy {
  static readonly ROOT_PRIORITY = 1000.0

  constructor(private readonly k: number = 32.0) {
    if (!Number.isFinite(k) || k <= 0) {
      throw new Error("Elo k must be a positive finite number")
    }
  }

  seed(parent: Hypothesis | null): number {
    return parent === null ? EloPolicy.ROOT_PRIORITY : parent.priority
  }

  priority(hypothesis: Hypothesis, _tree: ResearchTree): number {
    return hypothesis.priority
  }

  settle(referencePriority: number, outcome: Outcome): number {
    if (!Number.isFinite(referencePriority)) {
      throw new Error("reference priority must be finite")
    }
    const score =
      outcome === Outcome.WIN ? 1.0 : outcome === Outcome.DRAW ? 0.5 : 0.0
    return referencePriority + this.k * (score - 0.5)
  }
}
